---
title: "TCGA Playground"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 3
---


# Download data


"Harmonized data" will be downloaded from **TCGA** as a `SummarizedExperiment` object:

More details on this object type can be seen [here](https://bioconductor.org/packages/devel/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html)

library(ggplot2)
library(tidyverse)

## RNAseq data

```{r, eval = F}

## Install TCGA if you don't have it (restart session after installing)

# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# 
# BiocManager::install("TCGAbiolinks")

library(TCGAbiolinks)
library(SummarizedExperiment)

# Gene expression aligned against hg38
query <- GDCquery(
  project = "TCGA-SKCM",
  data.category = "Transcriptome Profiling",
  data.type = "Gene Expression Quantification", 
  workflow.type = "STAR - Counts")


GDCdownload(query = query)

skcm_rna <- GDCprepare(query = query)

saveRDS(skcm_rna, "skcm_rna.rds")

colnames(colData(skcm_rna))
table(skcm_rna@colData$vital_status)

```


## miRNAseq data

```{r, eval=F}

query <- GDCquery(
  project = "TCGA-SKCM",
  data.category = "Transcriptome Profiling",
  data.type = "miRNA Expression Quantification")


GDCdownload(query = query)

skcm_mir <- GDCprepare(query = query)

saveRDS(skcm_mir, "skcm_mir.rds")

```

## Subtype information

```{r, eval=F}

skcm_subtype <- TCGAquery_subtype("SKCM")

saveRDS(skcm_subtype, "skcm_subtype.rds")

```




# Explore data


Familiarize yourself with the data sets

- Samples are identified by barcodes. See details [here](https://docs.gdc.cancer.gov/Encyclopedia/pages/TCGA_Barcode/)
##RNAseq extract

```{r}

skcm_rna <- readRDS("skcm_rna.rds")

# Extract RNAseq data
rna <- assay(skcm_rna)


```


##Gene info
```{r}


gene_data <- as.data.frame(rowData(skcm_rna))

gene_data[1:10, 1:10]


anyDuplicated(gene_data$gene_id) #0 means, they are unique observations

anyDuplicated(gene_data$gene_name) 

sum(duplicated(gene_data$gene_name)) #total duplication

dup_genes <- duplicated(gene_data$gene_name)| duplicated(gene_data$gene_name, fromLast=F)

View(gene_data[dup_genes, ])



# Use this information for gene renaming


```

###Duplications
```{r}

# dealing with duplicate gene ids

gene_data <- gene_data %>% 
  mutate(duplicated_name = duplicated(gene_name) | duplicated(gene_name, fromLast=T)) %>% 
  mutate(new_name = case_when(
    duplicated_name == TRUE ~ paste(gene_name, gene_id, sep = "_"),
    duplicated_name == FALSE ~ gene_name,
    TRUE ~ "XXXXXXXXXXXXXXXXXXXXXXXXXXXX"
  )) %>% 
  mutate(new_name = make.names(new_name)) #R friendly syntax


anyDuplicated(gene_data$gene_name)
anyDuplicated(gene_data$new_name)

```



##Patient data extraction
```{r}

patient_data <- as.data.frame(colData(skcm_rna))

anyDuplicated(patient_data$patient)

which(duplicated(patient_data$patient))

which(duplicated(patient_data$patient) | duplicated(patient_data$patient, fromLast=T))

dup_patients <- which(duplicated(patient_data$patient) | duplicated(patient_data$patient, fromLast=T))


```




#Normalization

##miRNA
```{r}


head(colnames(skcm_mir))

#select columns that contain "per million", value=T gives values
sel_cols <- c(grep(pattern = "reads_per_million",
                   x = colnames(skcm_mir), 
                   ignore.case = T, value = T))


head(sel_cols)


goodmir <- skcm_mir[, sel_cols]

#make linear values logarithmic by apply function. margin indicates if it is for col or row.

goodmir <- apply(goodmir, MARGIN = 2, FUN = function(x){log(x+1,base = 10)})

dim(skcm_mir)

dim(goodmir) #only 452 per million data

#miRNA names are lost in goodmir

#to get rid of a specific string
colnames(goodmir) <- gsub(pattern = "reads_per_million_miRNA_mapped_",
                          replacement = "",
                          x = colnames(goodmir))

#transform
t_mir <- goodmir %>% 
  t() %>% 
  as.data.frame()

#give names to columns
colnames(t_mir) <- make.names(skcm_mir$miRNA_ID)


#add barcode

t_mir <- t_mir %>% 
  mutate(sample_ID = substr(rownames(.),1,16)) %>% 
  select(sample_ID, everything())



```


##RNA normalization
```{r}

library(dplyr)

norm_rna <- edgeR::cpm(rna) 
#count per million, normalization

norm_rna <- log(norm_rna+1, base = 10)
identical(rownames(norm_rna), gene_data$gene_id)

rownames(norm_rna) <- gene_data$new_name

```


### RNA transformation & adding immune signature to transformed RNA data frame 
```{r}
#move sample ids from cols to rows
t_rna <- data.frame(t(norm_rna))

#calculate average immune signature of specific genes
immune <- c("CD8A", "CD3E", "GZMB", "PRF1")

immune_sign <- rowMeans(select(t_rna, immune))

t_rna <- t_rna %>% 
  mutate(immune_signature = immune_sign) %>% 
  select(immune_signature, everything())

# extract patient barcode and add as a column, barcode is first 12 of id

t_rna <- t_rna %>% 
  mutate(sample_ID = substr(rownames(.), 1,16 )) %>% 
  select(sample_ID, everything())
 
```



###Survival
```{r}
clinical <- patient_data[c("patient", "vital_status", "days_to_death", "days_to_last_follow_up")]

#a new boolean varaible created 
clinical$deceased = clinical$vital_status == "Dead"

#a variable that shows overall survival. if the person is dead, it is equal to days_to_death, if alive, gives days_to_last_follow_up
clinical$overall_survival = ifelse(clinical$deceased,
                                   clinical$days_to_death,
                                   clinical$days_to_last_follow_up)

#add immune_signature varaible to a new clinical data frame
clin_immune <- clinical %>% 
  mutate(immune_signature = t_rna$immune_signature)


#Calculating median of immune signature
med_immune <- median(t_rna$immune_signature)


#classify whether the immune signature is lower or higher than the median
clin_immune <- clin_immune %>% 
  mutate(immune_median = case_when(
    immune_signature <= med_immune ~ "Low",
    immune_signature > med_immune ~ "High"
  ))
```


#Human-readable metadata

```{r}

#Putting "Not Available" in subtype data
#Correcting wildtype indications
#All values are classified as factor

for (i in 1:112) {
  levels(skcm_subtype[,i])[which(levels(skcm_subtype[,i]) == "-")] <- "Not Available"
  levels(skcm_subtype[,i])[which(levels(skcm_subtype[,i]) == "[Not Available]")] <- "Not Available"
  levels(skcm_subtype[,i])[which(levels(skcm_subtype[,i]) == "[Not Applicable]")] <- "Not Available"
  levels(skcm_subtype[,i])[which(levels(skcm_subtype[,i]) == "n/a")] <- "Not Available"
  levels(skcm_subtype[,i])[which(levels(skcm_subtype[,i]) == "NA")] <- "Not Available"
  
  
  levels(skcm_subtype[,i])[which(levels(skcm_subtype[,i]) == "wt")] <- "Wildtype"
  levels(skcm_subtype[,i])[which(levels(skcm_subtype[,i]) == "WT")] <- "Wildtype"

}


#Putting "Not Available" in subtype data
#Correcting wildtype indications

for (i in 1:173) {
  if (is.factor(pat1[,i])) {
    levels(pat1[,i])[which(levels(pat1[,i]) == "-")] <- "Not Available"
    levels(pat1[,i])[which(levels(pat1[,i]) == "n/a")] <- "Not Available"
    
    levels(pat1[,i])[which(levels(pat1[,i]) == "wt")] <- "Wildtype"
  }
  

}


```


# Leukocyte fraction and CIBERSORT
##Leukocyte fraction
```{r}
library(stringr)

leuk_data <- read.csv("TCGA_all_leuk_estimate.masked.20170107.tsv", sep="\t", header=TRUE)

#the data do not have column names, so i gave accordingly
names(leuk_data) <- c("Project", "Sample", "Leukocyte")

leuk_skcm <- leuk_data[leuk_data$Project=="SKCM",]

#leukocyte fractions of patients 
leuk_skcm <- leuk_skcm %>% 
  mutate(patient = substr(Sample, 1,12)) %>% 
  select(patient, Leukocyte)

leuk_skcm$patient <- gsub("[.]", "-", leuk_skcm$patient)
```

##CIBERSORT
```{r}
ciber_data <- read.csv("TCGA.Kallisto.fullIDs.cibersort.relative.tsv", sep="\t", header=TRUE)

ciber_skcm <- ciber_data[ciber_data$CancerType=="SKCM",]

#leukocyte fractions of patients 
ciber_skcm <- ciber_skcm %>% 
  mutate(patient = substr(SampleID, 1,12)) %>% 
  relocate(patient, .after = SampleID)

ciber_skcm$patient <- gsub("[.]", "-", ciber_skcm$patient)
ciber_skcm$SampleID <- gsub("[.]", "-", ciber_skcm$SampleID)

ciber_skcm$CancerType <- NULL #unnecessary info, we know that we work with SKCM 


```


# Merging
##RNA & miRNA
```{r}
#Since t_rna data is huge I relocated the specified genes, (CD8A, CD3E, GZMB, PRF1, FOXP3) to the beginning and used the first 100 columns for merging.
t_rna <- t_rna %>% 
  relocate(CD8A, .before = TSPAN6)

t_rna <- t_rna %>% 
  relocate(CD3E, .before = TSPAN6)

t_rna <- t_rna %>% 
  relocate(GZMB, .before = TSPAN6)

t_rna <- t_rna %>% 
  relocate(PRF1, .before = TSPAN6)

t_rna <- t_rna %>% 
  relocate(FOXP3, .before = TSPAN6)

t_rna <- t_rna %>% 
  relocate(immune_signature, .after = sample_ID)

#hsa.mir.155.5p doesnt exist in t_mir
t_mir <- t_mir %>% 
  relocate(hsa.mir.155, .before= hsa.let.7a.1)


#RNA and miRNA data merged by sample id

merged_id <- full_join(t_rna[,1:100], t_mir[, 1:100], by = "sample_ID")

#a column of patient names are generated
merged_2 <- merged_id %>% 
  mutate(patient = substr(sample_ID, 1,12)) %>% 
  select(patient, everything())
```

## add subtype
```{r}
#subtype and merged RNA&miRNA data are joined
merged_sub <- full_join(merged_2, skcm_subtype, by="patient")

###some patients that are present in rna&mirna data are not found in subtype. that's why it generates NAs.
```

## add leukocyte fraction and CIBERSORT

```{r}
merged_leuk <- full_join(merged_sub, leuk_skcm, by= "patient")

merged_final <- full_join(merged_leuk, ciber_skcm, by="patient")

merged_final$SampleID <- NULL #the sampleID data coming from CIBERSORT is unnecessary since we already have it
```

#Cleaning patient info that has NA in subtype info in merged_final
```{r}
final <- merged_final

final <- na.omit(final)

```


#Plotting

##Boxplot

### CD8A in two gender classes
```{r}
merged_final %>% 
  filter(GENDER!="Not Available") %>% 
  ggplot(aes(GENDER, CD8A)) +
  geom_boxplot(na.rm=TRUE) +
  geom_jitter(aes(colour = GENDER)) +
  ggtitle("CD8A Expression by Gender") + xlab("Gender")

```
### immune signature vs FOXP3
```{r}
merged_final %>% 
  ggplot(aes(immune_signature, FOXP3)) +
  geom_boxplot(na.rm=TRUE) +
  geom_jitter(aes(colour = immune_signature)) +
  xlab("Immune Signature") + ggtitle("Immune Signature vs FOXP3 Expression Plot")
```


## Scatterplot

### hsa.mir.155.5p vs CD8A 
```{r}
#hsa.mir.155.5p is not found in t_mir

merged_final %>% 
  ggplot(aes(hsa.mir.155, CD8A)) +
  geom_point(na.rm=TRUE) +
  geom_smooth(method = "lm", se=TRUE) + #straight line with confidence intervals
  ggtitle("hsa.mir.155 vs CD8A Expression Plot")

```
### CD3E vs CD8A
```{r}
merged_final %>% 
  ggplot(aes(CD3E, CD8A)) +
  geom_point(na.rm=TRUE) +
  geom_smooth(method = "lm", se=TRUE) + #straight line with confidence intervals
  ggtitle("CD3E vs CD8A Expression Plot")
```


##Kaplan Meier curve
```{r}
#fit a survival model
library(survival)

fit = survfit(Surv(overall_survival, deceased) ~ immune_median, data=clin_immune)


library("survminer")
ggsurvplot(fit, data=clin_immune, pval=T, risk.table = T)


```


#Playground
```{r, fig.width=8, fig.height=}

boxplot(rna[,1:10])

boxplot(log1p(rna[,1:10])) #1 ekleyip log alıyor



norm_rna <- edgeR::cpm(rna) 
#count per million, normalizasyon için MANUEL

norm_rna <- log(norm_rna+1, base = 10)

boxplot(norm_rna[, 1:10])

dim(norm_rna)



# norm_rna2 <- edgeR::cpm(rna, log = TRUE, prior.count = 1) 
# 
# i <- 10
# 
# plot(norm_rna[,i], norm_rna2[,i]) 1 ve 2deki hastanın genlerini karşılaştırma


```

##Bar plot
```{r}

## Bar plot without NA values 
merged_final %>%
   filter(RNASEQ.CLUSTER_CONSENHIER!="-") %>%
     ggplot(aes(sample_ID, FGR, fill=RNASEQ.CLUSTER_CONSENHIER)) + 
        geom_col() 
```

##Density
```{r}

merged_final %>% 
  filter(ALL_PRIMARY_VS_METASTATIC!="-") %>% 
  ggplot(aes(x=FGR, fill=ALL_PRIMARY_VS_METASTATIC)) +
  geom_density(alpha=0.3)
```

##22.07.22
```{r}
"""
mutate(case_when(
  !duplicated -> olduğu gibi bırak
  duplicated -> gene_ensid
))
"""


```


# Session Info
```{r}
sessionInfo()

```


